substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVSymbolSectionImpl.java

```diff
@@ -0,0 +1,188 @@
+        int pos = 0;
+        enableLog(debugContext);
+        log(debugContext, "CVSymbolSectionImpl.createContent() adding records");
+        addRecords(debugContext);
+        log(debugContext, "CVSymbolSectionImpl.createContent() start");
+        /* Add header size. */
+        pos += Integer.BYTES;
+        /* Add sum of all record sizes. */
+        for (CVSymbolRecord record : cvRecords) {
+            pos = CVUtil.align4(pos);
+            pos = record.computeFullSize(pos);
+        }
+        /* Create a buffer that holds it all. */
+        byte[] buffer = new byte[pos];
+        super.setContent(buffer);
+        log(debugContext, "CVSymbolSectionImpl.createContent() end");
+    }
+
+    @Override
+    public void writeContent(DebugContext debugContext) {
+        int pos = 0;
+        enableLog(debugContext);
+        log(debugContext, "CVSymbolSectionImpl.writeContent() start recordcount=%d", cvRecords.size());
+        byte[] buffer = getContent();
+        /* Write section header. */
+        log(debugContext, "  [0x%08x] CV_SIGNATURE_C13", pos);
+        pos = CVUtil.putInt(CV_SIGNATURE_C13, buffer, pos);
+        /* Write all records. */
+        for (CVSymbolRecord record : cvRecords) {
+            pos = CVUtil.align4(pos);
+            log(debugContext, "  [0x%08x] %s", pos, record.toString());
+            record.logContents(debugContext);
+            pos = record.computeFullContents(buffer, pos);
+        }
+        log(debugContext, "CVSymbolSectionImpl.writeContent() end");
+    }
+
+    private void addRecords(DebugContext debugContext) {
+        addPrologueRecord();
+        addFunctionRecords(debugContext);
+        addFileRecord();
+        addStringTableRecord();
+    }
+
+    private void addPrologueRecord() {
+        CVSymbolSubsection prologue = new CVSymbolSubsection(cvDebugInfo);
+        CVSymbolSubrecord.CVObjectNameRecord objectNameRecord = new CVSymbolSubrecord.CVObjectNameRecord(cvDebugInfo);
+        if (objectNameRecord.isValid()) {
+            prologue.addRecord(objectNameRecord);
+        }
+        prologue.addRecord(new CVSymbolSubrecord.CVCompile3Record(cvDebugInfo));
+        prologue.addRecord(new CVSymbolSubrecord.CVEnvBlockRecord(cvDebugInfo));
+        addRecord(prologue);
+    }
+
+    private void addFunctionRecords(DebugContext debugContext) {
+        /* This will build and add many records for each function. */
+        new CVSymbolSubsectionBuilder(cvDebugInfo).build(debugContext);
+    }
+
+    private void addFileRecord() {
+        /* Files are added to this record during function record building. */
+        addRecord(fileTableRecord);
+    }
+
+    CVFileTableRecord getFileTableRecord() {
+        return this.fileTableRecord;
+    }
+
+    private void addStringTableRecord() {
+        CVSymbolRecord stringTableRecord = new CVStringTableRecord(cvDebugInfo, stringTable);
+        addRecord(stringTableRecord);
+    }
+
+    static final class CVStringTable {
```

Comment:
This is more of an observation, and while I think it would be good to address this, it’s okay to do it in a follow-up PR as well.

I find that `CVStringTable` and `CVFileTable` are conceptually similar enough that their implementations and use in code should follow similar patterns. However, this is not the case at the moment.

`CVStringTable` is a separate class that is passed around directly, and `CVStringTableRecord` is merely a wrapper. On the other hand, `CVFileTableRecord` is both a table and a record (there is no separate class `CVFileTable`), which is accessed via `CVDebugInfo`. In principle, I don't mind either approach, but what I don't like is that we have both.